Fixed Cesium rendering when clustering is turned on and the canvas is…

bdeining opened the pull request:
#### What does this PR do?

This PR uses billboards and dynamically created SVGs to render clustering markers.   This was done to avoid rendering issues, as well as provide better looking markers.
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?

@andrewkfiedler  @djblue 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).

@pklinef
@kcwire 
#### How should this be tested?

Build the Catalog UI, ingest data, perform a query, turn clustering on and switch workspaces
#### Any background context you want to provide?
#### What are the relevant tickets?
#### Screenshots (if appropriate)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

… hidden

andrewkfiedler committed:
Initializing catalog ui search

djblue committed:
Added catalog ui as a new feature as part of the search-ui-app.

andrewkfiedler committed:
Updates 2d map to resize correctly, updates results to appear on map

djblue committed:
Added some more stuff to the workspaces endpoint. Still need to work on update.

djblue committed:
Added update endpoint.

djblue committed:
Updated workspaces enpoint and added unit tests.

andrewkfiedler committed:
Adds seperated results, version of tab change (WIP)

andrewkfiedler committed:
Fixes up tab styling

andrewkfiedler committed:
Updating tab styling

andrewkfiedler committed:
Reverts tab styling to previous style

andrewkfiedler committed:
Adds selectability to results list, syncs selection between results list and map

andrewkfiedler committed:
Updates metacard views

andrewkfiedler committed:
Splits single metacard view from multi metacard view

andrewkfiedler committed:
Fixes issues with queries not updating their cql upon save

andrewkfiedler committed:
Updates debounce for changes on the central model (selected results, queries)

djblue committed:
Added policy plugin to prevent users with incorrect roles to access workspaces.

djblue committed:
Added policy plugin to prevent users with incorrect roles to access workspaces.

andrewkfiedler committed:
Adds a metacard basic and advanced view

andrewkfiedler committed:
Adds a thumbnail input

andrewkfiedler committed:
Adds new datetimepicker
 - Query will use this from now on.
 - New input that is based off this.

djblue committed:
Fixed policy plugin by not returning null.

Still need to figure out how to OR roles instead of AND.

andrewkfiedler committed:
Updates datepicker format
 - Added fractional seconds (added by manually clicking and typing)
 - Updates revert functionality to work with date inputs

andrewkfiedler committed:
Updates workspaces to use the new rest endpoint

andrewkfiedler committed:
Workspaces can be edited

andrewkfiedler committed:
Updates views to work with header and footer

djblue committed:
Working on adding roles to workspaces.

andrewkfiedler committed:
Updated backbone cometd to use amd format.

andrewkfiedler committed:
Fixes some jshint errors

andrewkfiedler committed:
Removes unused code

andrewkfiedler committed:
Updates loading screen, fixes it to disappear only after all loading has finished

andrewkfiedler committed:
Deletes unused code

andrewkfiedler committed:
Deletes unused code

djblue committed:
Still working on roles in workspaces.

rzwiefel committed:
Endpoints for associations and metacard edit + validation

rzwiefel committed:
Updated associations endpoint and Full endpoint to do what andrew wants it to

rzwiefel committed:
refactor endpoints and fix the put bug for metacard edit endpoint

rzwiefel committed:
Added correct return types to endpoints that were missing them

djblue committed:
Added itests for workspace endpoint.

rzwiefel committed:
updated metacard validation return object and added nullchecks everywhere

rzwiefel committed:
Fixed the validation object to look like Andrew wants it to

rzwiefel committed:
Refactored the validation response so that its not garbage code

rzwiefel committed:
Added default objects and empty arrays to be included in json object

rzwiefel committed:
Refactored endpoints some more and formatted files

rzwiefel committed:
Removed erroneous a

rzwiefel committed:
Changed history endpoint query to start index of 1

rzwiefel committed:
First part of associations endpoint redo

djblue committed:
Added security for workspaces.

rzwiefel committed:
Second part of Associations reimplementation

andrewkfiedler committed:
Adds metacard validity, quality report, history, associations.

Adds a new menu (file, edit, view, help . . .), with a few of the menus populated.

Adds a new WIP workspace home view.

Adds rerouting for deleted workspaces, or workspaces that users don't have access to.

andrewkfiedler committed:
Updates lightbox styling

andrewkfiedler committed:
Adds ingest view, recently uploaded view, preferences view, adds popouts

andrewkfiedler committed:
Updates workspace query collection to properly unassign colors
 - Changes `query.getId` to `query.getId()` so that colors are properly unassigned and then reassigned upon sync.

andrewkfiedler committed:
Adds id attribute to queries
 - Fixes issues with backbone syncing of workspaces, where it would repopulate the query collection and cause all sorts of issues.

andrewkfiedler committed:
Updates query saving to work with new query ids
 - Previously we copied all cloned query attributes into the saved query, but if we do that now it'll overwrite the original's id.  Hence, we need to get a copy and then delete the id, and then set the modified attributes in the original query.

andrewkfiedler committed:
Adds bulk editing of metacards

djblue committed:
Mapped workspace metacards to related.

andrewkfiedler committed:
Adds media queries to workspace home view for smaller screens

andrewkfiedler committed:
Adds expanded view to templates

andrewkfiedler committed:
Fixes select height in chrome

andrewkfiedler committed:
Updates templates view to unexpand upon choosing a template

djblue committed:
Updated configuration for new search ui.

rzwiefel committed:
Full working batch metacard edit endpoint (RCZ- DeleteRequestImpl make full ddf PR)

andrewkfiedler committed:
Adds dropdowns to workspaces home screen
 - Also moves controls down to the workspace items rather than in the banner so they are near what they affect.

andrewkfiedler committed:
Adds changing of workspace title

djblue committed:
Replaced edn config with json.

rzwiefel committed:
Updated History endpoint to return the updated metacard on a revert

andrewkfiedler committed:
Adds editing capability

djblue committed:
Added default imagery provider.

djblue committed:
More configuration changes.

- Updated metatype descriptions to display json.
- Added better error handling for configuration.
- Updated catalog ui search to use new config endpoint.

djblue committed:
Added new configuration for ui.

djblue committed:
Removed edn-java as a maven dependency.

djblue committed:
Updated things so we can run a full build.

- Made some inner classes static.
- Added some find bug suppressions.
- Reduced code coverage requirements.
- Enabled checkstyles.
- Added some licence headers.
- Removed some unused imports.

djblue committed:
Updated configuration to use its own unique endpoint.

djblue committed:
Fixed proxied map tiles.

djblue committed:
Ported over login view.

rzwiefel committed:
Added batch get endpoint + refactored some of endpoints

rzwiefel committed:
Changed the metacard type format to be a map instead of list

rzwiefel committed:
Refactored EndpointUtil and RealEndpoint

pklinef committed:
Replaced CometD search with AJAX search

pklinef committed:
AJAX query service cleanup

pklinef committed:
Added query status tab

pklinef committed:
Switched from cql to query title

pklinef committed:
Add CQL parsing so queries in workspaces can be restored

pklinef committed:
Added a status message if a search is active in result selector

pklinef committed:
Added multi-selection to maps and replaced wreqr with store models

pklinef committed:
Added pending and error icons to status in result selector

djblue committed:
Added local storage persistence for guest users.

andrewkfiedler committed:
Adds multivalue / bulk attribute display & edit

andrewkfiedler committed:
Adds metacard actions view

andrewkfiedler committed:
Adds metacard archive view

andrewkfiedler committed:
Adds confirmation to archiving

djblue committed:
Added login form to home-menu.

djblue committed:
Added more configuration options.

- attributeAliases is for changing how attributes are displayed.
- hiddenAttributes is for keeping attribute from being displayed.

djblue committed:
Updated some things so we can do a full build.

pklinef committed:
Removed extra AJAX request when displaying result summary and details

pklinef committed:
Added paging and removed nested events to improve rendering performance

pklinef committed:
Reduced size of paging controls

andrewkfiedler committed:
Adds advanced search, query actions
 - Advanced search can now be performed using the filter builder.
 - Query actions, such as duplicate, delete, and run are now available from a menu in the query item.
 - Added a query item card to encapsulate representation of queries.

andrewkfiedler committed:
Updates metacard-associations
 - There was an issue with toJSON on queries not including the result object.

andrewkfiedler committed:
Removes some dead code, fixes saving in metacard views / dates

andrewkfiedler committed:
Updates empty query message to display properly
 - With the change to make query items their own view, the query-selector no longer constantly rerenders.  As such, this logic has been moved to work with the new render once paradigm.

andrewkfiedler committed:
Updates query model to not use DURING operator
 - The query model now uses a before and an after to accomplish a during.
 - Updated comparator view to be compatible with casing of cql operators.

pklinef committed:
Fixed query defaults like title from being shared with new queries

pklinef committed:
Fixed query result count updates

pklinef committed:
Fixed parsing of double quoted property names

andrewkfiedler committed:
Updates properties to be put in double quotes
 - Double quotes are necessary for the cql parser

andrewkfiedler committed:
Adds editing mode to filter-builder, filter, and dropdowns

andrewkfiedler committed:
Updates filters to guard against bad input
 - Makes it impossible to delete the last filter, so we no longer get cql consisting of empty parentheses.

andrewkfiedler committed:
Filter and filter-builders now properly hide and show delete buttons

pklinef committed:
Added padding to query status view

andrewkfiedler committed:
Updates content view to stop rerendering panel two
 - Rerendering is now delegated to the tabs-query.view
 - Also fixes a view destroyed issue

andrewkfiedler committed:
Updates popout panel to rerender self unless necessary
 - Allows switching between metacards and staying on the same tab.
 - Popout panel now closes as you go between queries, results, and saved items

andrewkfiedler committed:
Updates query-status view to only listen if the query exists
 - Previously we were attempting to listen to undefined.

pklinef committed:
Added query status of last time results have been updated

pklinef committed:
Made query actions icon larger

andrewkfiedler committed:
Adds query-settings view
 - Allows editing title, sorting, and sources for a query.
 - Also updates query-advanced and settings to be editing when viewing a new query by default.  Canceling will close the popout (if it's a new query).

djblue committed:
Ported over notification view.

pklinef committed:
Improved spacing of content buttons in filter builder

andrewkfiedler committed:
Updates maps to pan to selection
 - Map will pan (without zooming) to selected results (if there is only one selected result).  Selecting multiple results will not result in panning, so that the user is not disoriented.

djblue committed:
Fixed jshint style issue.

pklinef committed:
Added missing semicolon

andrewkfiedler committed:
Fixes issue with null dates
 - Null dates were previously displayed inproperly, and changes weren't detected properly.
 - Also updates history and advanced views to propagate responses with attributes to their respective metacards.

andrewkfiedler committed:
Adds archiving view for when multiple metacards are selected

djblue committed:
Removed color preferences tab.

andrewkfiedler committed:
Updates lightbox to not always show scrollbars
 - The previous commit changed overflow to scroll, but it should've been auto.

djblue committed:
Ported over tasks.

djblue committed:
Added configurable source polling interval.

andrewkfiedler committed:
Adds routing to individual metacards
 - Users can now route to individual metacards.  Breadcrumbs are dynamic based on whether or nut the user is currently in a workspace.
 - Pulled some of the listener logic out of cesium.controller and openlayers.controller.  These listeners are dependent on what the map is being used for, so they're now attached elsewhere (content.view for instance).
 - Updated cesium.controller and openlayers.controller to take in an element to put the map in instead of relying on a querySelector on id.  This will allow us to have multiple maps (in addition to having pulled out the listener logic).
 - Moved metacard interactions into a dropdown.  Dropdown options change based on whether the user is in a workspace, if the metacard has already been saved, etc.  Need to add a modal for selecting a workspace when saving from outside a workspace.
 - Expand metacard interaction now routes to the metacard.  So long as a user has permission, they can go to the link and view the metacard (even if they don't have a workspace).
 - Updated dropdown companion logic to adjust when there isn't proper room on the sides.
 - Updated dropdown companion styling to make it pop more against any background.  It has a white box shadow (small, like a border) and then a black box-shadow that gives it depth.
 - Removed dropdown companion tails for the time being.
 - Updates cesium.metacard and openlayers.metacard to only perform certain functions when the displayed object is in a result set from a query.  Lots of uncoupling still needs to be done to make it easier to have multiple maps in the future.  Certain listeners need to be moved out.
 - Updated thumbnail.view to restrict height to a max of 200px on uploaded images.
 - Added a navigation.view which should be rolled out to each page (deprecating the existing navigation bars).  It allows setting a middle view, but keeps the left and right side of the navigation consistent between pages.
 - Updated metacard-basic and metacard-advanced views to check all collections of metacards in each query in each workspace and update any that have the same id.  This is necessary to keep duplicate results in sync, as well as to facilitate easier routing to specific metacards.  If this becomes a performance issue (since we're checking every metacard we currently have in memory), we'll need to revisit how to keep things in sync.
 - Updated a few minor styles in filters.

andrewkfiedler committed:
Adds navigation to metacard associations view
 - Users can click on an associations and get routed to that particular metacard.

andrewkfiedler committed:
Updated styling for navigating association links

andrewkfiedler committed:
Updates results view to use a dropdown
 - Results view now uses a dropdown instead of tabs to switch between queries.
 - Switched away from using the menus for dropdowns.  Now the dropdown view is used.

andrewkfiedler committed:
Seperates status and results details to their own lines

andrewkfiedler committed:
Updates how dropdown companions are destroyed

pklinef committed:
Removed unused query menu component

pklinef committed:
Fix result final page count

pklinef committed:
Update status if query request fails

pklinef committed:
Make match counting optional

pklinef committed:
Improve term match counting performance

andrewkfiedler committed:
Updates loading to be constrained to specific view doing the loading

andrewkfiedler committed:
Removes dead code, console log
 - In addition, changes the bulk edit view for metacards to use the new localized loading.

rzwiefel committed:
Moved endpoints to Spark, updated UI to use new endpoints

rzwiefel committed:
Fixed associations bug when collection for associationType doesn't exist

andrewkfiedler committed:
Adds a factory for simple enum views

andrewkfiedler committed:
Adds a geo input view, fixes issues with CQL parser
 - CQL parser can now handle UNITS.
 - CQL parser can now correctly handle Dates (previously it did not return an ISO formatted date).
 - CQL parser no longer returns different writes / reads upon successive calls to write / read. - Updated the CQL.factory to not put erroneous spaces in the text it was generating for polygons.
 - Added sanitizeGeometryCQL to filter-build.view to allow the backend to correctly understand the cql that is being generated.  It could not handle quotes around POLYGONS or POINTS, which the cql parser generates (since they are seen as values).  In the future, we might be able to update the parser and remove this, or update the backend to allow quotes around these.
 - Added sanitizeForCQL for filter.view to allow the backend to correctly understand the cql being generated.  It could not handle '[' or ']' so these are replaced with '(' or ')' respectively.
 - Added methods for turning bbox to polygon in filter.view.  Also added a method to turn polygons into a form that cql can understand.
 - Added a method to handle geometry filters to filter.view.
 - Added location-old component, which takes a chunk from the old query view and abstracts it to handle all the location details.
 - Updated cesium.bbox.js, cesium.polygon.js, cesium.circle.js, openlayers.bbox.js, and openlayers.polygon.js to handle the concept of multiple views.

andrewkfiedler committed:
Adds way to close lightbox from child elements
 - Adds a way to trigger the close of a lightbox from a child element of the lightbox.
 - Removes some console logs.
 - Updates filter.view to handle geometry and location types as geometry filters.

andrewkfiedler committed:
Adds an interaction to stop a query
 - A query can be stopped while being run, but any results that have already come back will still be able to be accessed.

andrewkfiedler committed:
Adds indication of source availability
 - Sources that are unreachable will be greyed out, and have a warning icon next to them.

andrewkfiedler committed:
Updates UI to use the attribute aliases

andrewkfiedler committed:
Updates summary view for metacard to display configured summary view

djblue committed:
Updated catalog-ui-search to use branding-api.

pklinef committed:
Added location case to filter comparator view

andrewkfiedler committed:
Fixes validation serialization issue

djblue committed:
Login fixes.

- CometD endpoint was updated.
- Default roles for user model were set to ['guest'].
- Login dropdown uses method to check for guest.

andrewkfiedler committed:
Adds ability to download metacard(s)

djblue committed:
Updated itests.

djblue committed:
Added Workspace Sharing.

- Restructure files.
- Added the WorkspaceAccessPlugin to allow only the owner/system
  to update workspaces.
- Removed ability to configure public workspaces.
- Removed roles workspace metacard attribute.
- Added metacrd.sharing attribute for sharing by roles/email with
  permissions of view/edit.

djblue committed:
Added capability for guest users to interact with backend.
  - Locally created workspaces get tagged with { localStorage: true }
  - Tagged workspace get saved to localStorage, otherwise get saved to
    backend so both types can interoperate.

djblue committed:
Added sharing modal in ui.

djblue committed:
Updated grunt proxy to work with secure JSESSIONID cookies.

NOTE: you should only use the proxy or the endpoints directly not
      both, if you do your session will be invalidated for the other.

djblue committed:
Added endpoints for getting user info and saving preferences.

djblue committed:
Updated search ui to use the new user endpoints.

djblue committed:
Reverted UserService.

djblue committed:
Reverted DeleteRequest.

djblue committed:
Documented and updated sharing schema.

pklinef committed:
Made actions a list instead of a map due to duplicate action ids

andrewkfiedler committed:
Adds result cards, saved item view, gallery mode
 - Split out some of the logic in result-selector into more modular pieces, so each could rerender independently or be reused elsewhere (result-item.view and paging.view).
 - Added a saved item view.  For now it queries for all the ids that have been saved to the workspace.
 - Added a gallery mode to take advantage of the new result-item.view.  Now users can switch to a view that adds a thumbnail preview above the title in the result list.
 - Updated the dropdown.companion.view to be more performant.  It now waits until the first open to render and attach.
 - Split the query-item from the query-feed view, so the latter can rerender independtly whenever necessary.  This will prevent interference with other parts of the layout view.
 - Added starring and an actions menu beside result cards (see result-item.view).
 - Updated the routing to a metacard to use jquery.whenAll.  The normal jquery.when will resolve immediately if any of the requests fail, which is not the behavior we want.  The jquery.whenAll will wait for all of the requests to return.  In the future, it should be changed to use the first resolution that has a result.
 - Added two new actions to the file menu.  One to run all queries in the workspace, and one to stop all current queries.
 - Added a result indicator that will display the color of any query that contains the result.  The more queries that contain the result, the more colors will be split along the side of the result card.

andrewkfiedler committed:
Adds ability to duplicate a workspace
 - Duplication is now an action in the workspace action menu.

pklinef committed:
Removed debug console logging from openlayer bbox and polygon

pklinef committed:
Fixed thumbnails due to action map becoming a list

pklinef committed:
Updated default color list

andrewkfiedler committed:
Adds detection of deleted or unavailable metacards
 - User is now prompted with a friendly message.
 - If saved items are missing, the user has the option to remove them their workspace or keep and try again later.
 - If routed to a missing metacard, the user is redirected to the workspaces screen through a confirmation.

andrewkfiedler committed:
Updates routing metacard message
 - It now matches the prompt for missing saved items.

andrewkfiedler committed:
Removes unused code

andrewkfiedler committed:
Removes unused code

andrewkfiedler committed:
Adds client side filtering of results
 - Adds a filter view for setting a client side filter for results.
 - Updates dropdowns to handle being inside other dropdowns.

djblue committed:
Updated sharing policy.

- Actions are shared explicitly (read/update/delete)
- No actions are implied by other actions
  - update doesn't mean you can delete.
  - update doesn't mean you can read.
- WorkspaceSharingPolicyPlugin uses CollectionPermission constants for
  various action types.

djblue committed:
Ported over map overlays from standard search ui.

djblue committed:
Sharing ui updates.

- Removed workspace link in modal.
- Display sharing modal on File -> Share.
- Set default view permission for email based sharing.

pklinef committed:
Removed metacard sharing tab

pklinef committed:
Remove muted color from paging text

pklinef committed:
Use exact bower dependency versions

andrewkfiedler committed:
Updates result selector to listen to result sync events

andrewkfiedler committed:
Updates result selector handle paging / filtering

pklinef committed:
Change color order

pklinef committed:
Use current versions of versioning and persistence

andrewkfiedler committed:
Updates filter view to properly set default comparator
 - In the past it would always set a default, even if deserializing an existing filter (hence, it would override the deserialized value).  Now it will only set a default if the deserialized value does not make sense for the type.

pklinef committed:
Fix content type parsing in sources model

andrewkfiedler committed:
Adds client side sorting of results
 - Adds a result sort view to handle updating / resetting of the sorting.
 - Adds property to user model to remember user preference for sorting.

djblue committed:
Workspace and Preference updates.

- Workspace sorting/filtering/display are implemented.
- Workspace sorting/filtering/display are now preferences.
- Workspace preferences are auto saved on change.
- All user preferences are persisted.
- Guest preferences are persisted to local storage.
- User preferences are persisted to backend.

bdeining committed:
Added results clustering to Cesium Map and Toggle Button

bdeining committed:
Removed Comment and added a constant in Cesium Map Clustering

pklinef committed:
Remove assumption that resource uri can be used to get thumbnail uri

pklinef committed:
Save result display preference

pklinef committed:
Improve workspace input styling and remove unused elements

djblue committed:
Forgot to add dependency for Backbone when porting overlays.

djblue committed:
Added initial basic restore view.

djblue committed:
Updated archive wording.

rzwiefel committed:
Added Enumeration endpoint

andrewkfiedler committed:
Adds scheduling of queries and associated alerts
 - Queries now have a polling attribute that can be used to start polling that query.  It looks for new or modified results between the polling frequency chosen.
 - Users now have a preference for if alerts should persist between refreshes (sessions).  If they do persist, the user has to select a time after which an alert should expire and be automatically deleted.
 - Alerts are now part of the menu bar.  Clicking an alert will route the user to the specified alert.
 - Alerts are now a route.  If the alert exists, it will show the results in the left pane and allow everything possible from the normal results list (filtering, sorting, etc.)
 - If the alert doesn't exist, it redirects to the home page (workspaces).
 - The alert page shows a breadcrumb that reveals the associated workspace and query if still available.  If not, unknown is displayed.
 - Updates views to be agnostic regarding the selection interface they're interacting with, hence the maps, result list, etc. work independently between the normal workspace view and an alert view.
 - Updates the lightbox css to better center the title.
 - Adds a radio input, with a simple way to enact new views.

andrewkfiedler committed:
Adds an empty view for the alert-item.collection.view
 - Displays "No Recent Alerts"

andrewkfiedler committed:
Moves the syncing of active search results to the result-item.collection
 - Whatever the result list is currently displaying, the visualizations (map) will also display.

rzwiefel committed:
Added recent uploads logic

andrewkfiedler committed:
Updates default alert expiration to be 1 day

andrewkfiedler committed:
Adds support for nested filtering, blacklisting metacards
 - Users can now blacklist metacards from future results.  So long as they haven't left the screen where they blacklisted, they can unblacklist results as well.  They can also route directly to a metacard and unblacklist from there.
 - If the user wants to reset their blacklist, they can go to Edit and click reset blacklist.
 - The blacklist is tacked on to the user's filter, so support for nested clientside filtering is now available.  In the future, we might enable this to allow for more advanced clientside filtering (instead of having users adjust their queries).

andrewkfiedler committed:
Adds admin ability to customize the result preview
 - By configuring the result preview configuration, different attributes can be shown in the reresult item preview.

andrewkfiedler committed:
Adds enum inputs, fixes issues with saving multivalued attributes
 - If there is a validation enum defined for an attribute, the normal input will be replaced by an enum input.  This works for multivalued attributes, as well as bulk editing.
 - Fixes issues with saving multivalued attributes.  Before, the response from the patch was used directly on the advanced metacard view and the bulk basic metacard view.  They were updated to be more like the basic metacard view, which has a special case for building the attribute map when it comes to multivalued attributes.  This issue only appeared after saving a metacard (refreshing would fix the issue).

andrewkfiedler committed:
Adds enum inputs, fixes issues with saving multivalued attributes
 - If there is a validation enum defined for an attribute, the normal input will be replaced by an enum input.  This works for multivalued attributes, as well as bulk editing.
 - Fixes issues with saving multivalued attributes.  Before, the response from the patch was used directly on the advanced metacard view and the bulk basic metacard view.  They were updated to be more like the basic metacard view, which has a special case for building the attribute map when it comes to multivalued attributes.  This issue only appeared after saving a metacard (refreshing would fix the issue).

andrewkfiedler committed:
Updates single metacard edit / view to show validation issues

andrewkfiedler committed:
Updates bulk input to use simple dropdown instead of select

andrewkfiedler committed:
Updates bulk view to show validation next to each possible value
 - For now the logic lives in property.view, but it should be delegated down to the actual bulk input at some point.

andrewkfiedler committed:
Updates save and unsave to correctly show & update
 - Previously, save and unsave were shown in the result preview regardless of whether the user was in a workspace.
 - Now you can go to a workspace, then come back and the save buttons will correctly appear and be in sync with the saved metacards.

andrewkfiedler committed:
Adds recent uploads view as well as route to that view
 - Users can now route to their recent uploads.
 - Nothing will show up if the user isn't logged in.
 - In the future, we'll need update the recent uploads page to take advantage of the paging on the backend.  For now, the number of results they can view in their recent uploads is limited.

andrewkfiedler committed:
Adds duplicate grouping, fixes issues with drawing on multiple maps

 - Results are now deduped and grouped together.

 - The drawing functions for cesium and openlayers now only activate on the visible map.  Previously opening more than one map would break the location inputs.

 - Updates activeSearchResults to not be a pageable collection.  There were issues with events sticking around after the view was destroyed.

djblue commented on catalog/ui/catalog-ui-search/src/main/webapp/js/widgets/cesium.mapclustering.js:
```diff
@@ -174,11 +170,21 @@ define(['backbone',
             return viewer.camera.pickEllipsoid(new Cesium.Cartesian2(width * 60 / 100, height / 2));
+        },
+        getSvgImage : function(value) {
+            var svgString = '<svg xmlns="http://www.w3.org/2000/svg" width="25" height="25">';
```
We might want to use a handlebars template here. Something like: 

``` javascript
var svg = hbs.compile(template)({
  fill: value.color.toCssColorString(),
  count: value.points.length
});

return window.btoa('data:image/svg+xml;base64,' + svg);
```

would work.

bdeining replied:
Added a handlebars template

bdeining committed:
Fixed Cesium rendering when clustering is turned on and the canvas is hidden

andrewkfiedler committed:
Updates clustering toolbar button to correctly attach to map
 - Previously, it would attach multiple times to multiple maps.

andrewkfiedler commented:
👍 Worked fine on my machine.

~~I ran into an issue with the way variables are being used in cesium.mapclustering though.  There are multiple maps, so each MapClustering instance should maintain certain variables separately.~~

~~Specifically these (https://github.com/codice/ddf-ui/blob/master/catalog/ui/catalog-ui-search/src/main/webapp/js/widgets/cesium.mapclustering.js#L26):~~

```
    var clusters = [];

    var entities, viewer;
```

~~Not sure if you want to tackle that with this PR.  If not, I can push up a fix.~~

Nevermind, just looked at https://github.com/codice/ddf-ui/pull/6 where it looks like that's all addressed.

rzwiefel committed:
Fixed MetacardApplication after rebase

andrewkfiedler committed:
Fixes issue with rendering destroyed view
 - Metacard-title's parent view was rerendering on the same listener, so it was possible to cause a view already destroyed error.

andrewkfiedler committed:
Updates duplicate checking to only compare defined checksums
 - Previously, null checksums would be marked as the same or similar, causing them to get grouped as duplicates.  Now, if checksum is null it will be skipped as a possible comparator for duplicates / similarity.

andrewkfiedler committed:
Updates sorting to be done after duplicate grouping
 - The comparator was getting reset (since the duplicate result collection is a new collection), so the sorting would be off previously.  Sorting after insures that the user's defined sort is being utilized in the result list.

andrewkfiedler committed:
Fixes scrolling performance during bulk edit
 - The value list used to highjack the scrolling.  Now the value list has overflow set to hidden during editing.

andrewkfiedler committed:
Updates all navigation menus to be the same
 - Previously the navigation menus were different, now only the middle area will be different.
 - Deleted a bunch of old, now unused views.
 - Renamed a few files to make more sense.

andrewkfiedler committed:
Removes unused view

andrewkfiedler committed:
Updates query frequency for half hour to be half hour
 - It was accidentally set to 10 seconds for debugging purposes.

andrewkfiedler committed:
Removes some now unnecessary code
 - No longer need to wait for app since router will redirect for us.
 - A lot of methods in store were no longer being used, or were no longer needed due to the router redirect.

bdeining commented:
Rolled this into https://github.com/codice/ddf-ui/pull/6 for simplicity
